### DISCUSSION TITLE
Getting geometry stats from Model/Mesh

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#341] [Q&A] [answered]
I want to show some stats (nr of verts and triangles) for a loaded Model. I currently use this piece, but even with all the checks it leads to "Attempt to read/write protected memory" for certain models:

```
                    num_vertices = 0;
                    num_triangles = 0;

                    foreach (var node in model.Nodes)
                    {
                        Mesh mesh = node.Mesh;
                        if (mesh == null)
                            continue;
                        if (mesh.GetVerts() != null)                            
                            num_vertices += mesh.GetVerts().Length;
                        if (mesh.GetInds() != null)
                            num_triangles += mesh.GetInds().Length / 3;
                    }
                    Log.Info($"Model {num_vertices} vertices, {num_triangles} triangles");
```

Is this due to the Mesh arrays living in GPU memory? Of is there something else going on?

**Comment by maluoi** [comment, marked as answer]
`GetVerts` and `GetInds` are both pretty pricey calls that involve marshaling a pile of data up into C#! They should still be on CPU though, so it's not marshaling the data all the way up from the GPU at least. Not sure why it'd be crashing though! I can look into adding nicer getters to these.

You can also iterate through `model.Visuals` instead, all those `ModelNodes` are guaranteed to have valid `Mesh`es :) I haven't seen any crashes with this code so far!

```csharp
int vertCount = 0;
int triCount  = 0;

foreach (var node in model.Visuals)
{
	Mesh mesh = node.Mesh;
	vertCount += mesh.GetVerts().Length;
	triCount  += mesh.GetInds ().Length / 3;
}
Log.Info($"Model {vertCount} vertices, {triCount} triangles");
```

**Reply by maluoi** [reply]
I've added `Mesh.VertCount` and `Mesh.IndCount`, they'll be available in the next build! Should be faster and more reliable than `GetVerts().Length` and such!

**Reply by paulmelis** [reply]
Great!

**Comment by paulmelis** [comment]
With respect to `Model.Nodes` versus `Model.Visuals`, should I use the latter if I want to have accurate geometry bounds? Or put differently, is there a difference in the bounds they provide?

**Reply by maluoi** [reply]
The `Model` `Bounds` that StereoKit calculates are based on the `Mesh` `Bounds` corners in Model space. `ModelNode` locations aren't directly factored into that, other than to determine the `Mesh` transforms. (Using the corners here does mean you could probably get tighter bounds by looking at the combined model space verts of all the meshes!)

`Model.Nodes` and `Model.Visuals` are both views onto the same list, the difference is that `Nodes` is **_all_** the `ModelNodes` on the `Model`, where `Visuals` is **_all_** the nodes that have _**visible elements**_ (meshes) on them.

I hope that clears it up?
